### PR TITLE
Live reload sweep configs

### PR DESCRIPTION
**Pull request by fsamuel-bs**
**Goals (and why)**: AtlasDB customers should be able to live reload sweep configs, e.g. disabling sweep to reduce load on the Cassandra boxes.

**Implementation Description (bullets)**:
- Create an AtlasDbRuntimeConfig file to host live-reloadable config options.
- Add appropriate endpoints on TransactionManagers.

**Concerns (what feedback would you like?)**:
- Is this the correct approach?
- Are there any tests valuable to be added?

**Where should we start reviewing?**: TransactionManagers.

**Priority (whenever / two weeks / yesterday)**: Soon, since live-reload it's a somewhat high priority feature request.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1976)

### CODE REVIEW

**Comment by fsamuel-bs** [comment]
For a sample of what it takes to adapt this to existing products, check PR #2671 on internal core product.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java [hunk 1]
```diff
@@ -145,6 +147,21 @@ public static SerializableTransactionManager create(
...
      * Create a {@link SerializableTransactionManager} with provided configuration, a set of
+     * {@link Schema}s, and an environment in which to register HTTP server endpoints.
+     */
+    public static SerializableTransactionManager create(
+            AtlasDbConfig config,
+            java.util.function.Supplier<AtlasDbRuntimeConfig> runtimeConfig,
+            Set<Schema> schemas,
+            Environment env,
+            boolean allowHiddenTableAccess) {
+        log.info("Called TransactionManagers.create on thread {}. This should only happen once.",
```
Comment (nziebart) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java]: can we de-dup this log line with the one above?

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java [hunk 2]
```diff
@@ -174,6 +206,33 @@ private static SerializableTransactionManager create(
             LockServerOptions lockServerOptions,
             boolean allowHiddenTableAccess,
             String userAgent) {
+        AtlasDbRuntimeConfig runtimeConfig = ImmutableAtlasDbRuntimeConfig.builder()
```
Comment (nziebart) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java]: maybe pull this into a static factory method on `AtlasDBRuntimeConfig`

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java [hunk 3]
```diff
@@ -174,6 +206,33 @@ private static SerializableTransactionManager create(
...
+        return create(config,
+                () -> runtimeConfig,
+                schemas,
+                env,
+                lockServerOptions,
+                allowHiddenTableAccess,
+                userAgent);
+    }
+
+    private static SerializableTransactionManager create(
```
Comment (nziebart) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java]: can we consolidate the two private constructors into one?

File: atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbRuntimeConfig.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+     */
+    @Value.Default
+    public long getSweepPauseMillis() {
+        return AtlasDbConstants.DEFAULT_SWEEP_PAUSE_MILLIS;
+    }
+
+    /**
+     * The target number of (cell, timestamp) pairs to examine in a single run of the background sweeper.
+     */
+    // TODO(gbonik): make this Default after we delete the deprecated options. For now, we need to be able to detect
```
Comment (nziebart) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbRuntimeConfig.java]: let's take the opportunity to do these `TODO`s

Comment (gbonik) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbRuntimeConfig.java]: That would be a breaking change though.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java [hunk 5]
```diff
@@ -156,6 +173,21 @@ public static SerializableTransactionManager create(
         return create(config, schemas, env, lockServerOptions, allowHiddenTableAccess, UserAgents.DEFAULT_USER_AGENT);
     }
 
+    /**
+     * Create a {@link SerializableTransactionManager} with provided configuration, a set of
+     * {@link Schema}s, {@link LockServerOptions}, and an environment in which to register HTTP server endpoints.
+     */
+    public static SerializableTransactionManager create(
```
Comment (gbonik) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java]: The chain of various create() methods is now almost impossible to follow without an IDE :)

Do I understand correctly that this change doesn't actually implement live reloading and we still use a constant supplier?

File: atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbRuntimeConfig.java (hunk 4 above)
Comment (nziebart) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbRuntimeConfig.java]: This is a new class, so should be fine right?

Comment (gbonik) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbRuntimeConfig.java]: These fields still come from the corresponding config parameters which are user-facing. We replaced two old parameters (getSweepBatchSize, getSweepCellBatchSize) with three new ones (getSweepReadLimit, getSweepCandidateBatchHint, getSweepDeleteBatchHint). Currently, to avoid a hard user break, if the old parameters are explicitly set in the config file, we give a warning and derive reasonable values for the new parameters from the old ones

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java (hunk 5 above)
Comment (fsamuel-bs) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java]: Agreed that the chain of create() methods is awful. If there's any way of improving its readability, please let me know. :(

This change does implement live reloading if the callers use the new public method
```
    create(
            AtlasDbConfig config,
            java.util.function.Supplier<AtlasDbRuntimeConfig> runtimeConfig,
            Set<Schema> schemas,
            Environment env,
            boolean allowHiddenTableAccess)
```

Since we do not want to make a breaking change, for the callers of the old method we preserve the original behavior and create a constant supplier.

**Comment by fsamuel-bs** [comment]
I've duplicated our chain of `create()` methods API, since we're probably going to need the added methods to live-reload configs in other products. Even though I did this, it does not look good. Is there a better way of adding support to live-reloading configs without duplicating and deprecatings the old API?

File: atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbRuntimeConfig.java (hunk 4 above)
Comment (gsheasby) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbRuntimeConfig.java]: Yep, we don't want to delete the deprecated configs just yet.

Comment (fsamuel-bs) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbRuntimeConfig.java]: This is already breaking by adding a new config block, the `atlas-runtime`, which will be live-reloaded... I don't know if we want to not support getSweepBatchSize, getSweepCellBatchSize on the new config though. This would potentially make the upgrade to a new atlas version harder (even though not by a large margin).

Comment (gsheasby) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbRuntimeConfig.java]: Oh, now I see the whole list - yep, we have no need to copy over the old config settings to the new block.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java [hunk 6]
```diff
@@ -131,7 +131,10 @@ public int getBackgroundScrubBatchSize() {
      * have been overwritten or deleted. This differs from scrubbing
      * because it is an untargeted cleaning process that scans all data
      * looking for cells to delete.
+     * @deprecated Use {@link AtlasDbRuntimeConfig#enableSweep} to make this value
+     * live-reloadable.
```
Comment (gsheasby) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java]: We need to check (in Value.Check below) that these values are not also present in the RuntimeConfig block.

Comment (gsheasby) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java]: Given that the configs are created separately - not sure if this is possible. Perhaps some validation (and logging a warning if it fails) in `TransactionManagers.create`?

**Comment by fsamuel-bs** [comment]
There are a couple of things missing with this PR, as commented above:
- documentation: we should docs for the new config
- tests: there are no tests in this current PR. We could potentially add them on a separate PR though.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java [hunk 7]
```diff
@@ -235,6 +236,7 @@ public static SerializableTransactionManager create(
             LockServerOptions lockServerOptions,
             boolean allowHiddenTableAccess,
             Class<?> callingClass) {
+        validateRuntimeConfig(runtimeConfig.get());
```
Comment (gsheasby) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java]: nit: can make the guarantees easier to enforce by pushing the validation method down into `create`.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java [hunk 8]
```diff
@@ -342,27 +344,62 @@ private static SerializableTransactionManager create(
...
-    private static SweepBatchConfig getSweepBatchConfig(AtlasDbRuntimeConfig config) {
+    private static void validateRuntimeConfig(@Nullable AtlasDbRuntimeConfig runtimeConfig) {
+        if (runtimeConfig == null) {
+            log.warn("This service uses live reload configs. Please use the 'atlas-runtime' block for specifying"
+                    + " live-reloadable atlas configs. Default to using configs specified on the 'atlas' block");
+            return;
+        }
+
+        if (runtimeConfig.enableSweep() == null) {
+            log.warn("Use the 'enableSweep' config on the atlas-runtime block to make it live-reloadable");
```
Comment (gsheasby) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java]: In my opinion, these should only be flagged if the corresponding non-runtime config is set. In practice, we aim for most customers being fine with the defaults, meaning that these warnings would create a lot of noise.

**Comment by tpetracca** [comment]
In practice, I'd like to see this actually done in a manual test, using one of the platform's atlas services to actually do the live reloading before we merge.

File: docs/source/release_notes/release-notes.rst [hunk 9]
```diff
@@ -46,6 +46,32 @@ develop
...
+             - `getSweepReadLimit`
+             - `getSweepCandidateBatchHint`
+             - `getSweepDeleteBatchHint`
+
+           And the deprecated config options:
+
+             - `getSweepBatchSize`
+             - `getSweepCellBatchSize`
+
+           (`Pull Request <https://github.com/palantir/atlasdb/pull/1976>`__)
```
Comment (jeremyk-91) [inline docs/source/release_notes/release-notes.rst]: I think for the above it might be better to point users to the existing sweep config docs, saying that the options/way you specify it is essentially the same.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java [hunk 10]
```diff
@@ -144,31 +164,86 @@ public static SerializableTransactionManager create(
...
             AtlasDbConfig config,
+            java.util.function.Supplier<AtlasDbRuntimeConfig> runtimeConfig,
             Set<Schema> schemas,
             Environment env,
             LockServerOptions lockServerOptions,
             boolean allowHiddenTableAccess,
             Class<?> callingClass) {
-        return create(config, schemas, env, lockServerOptions, allowHiddenTableAccess,
+        validateRuntimeConfig(runtimeConfig.get());
+        return create(config, runtimeConfig, schemas, env, lockServerOptions, allowHiddenTableAccess,
```
Comment (jeremyk-91) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java]: This looks like a check-then-act; the config validated here might not be the same as the one passed into the actual TM.create(). I realise this is fine for now since validate(...) only logs, but we should be aware of this in the event that validate() later on actually rejects configs for whatever reason.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java [hunk 11]
```diff
@@ -269,27 +344,62 @@ private static SerializableTransactionManager create(
...
     }
 
+    private static int chooseBestValue(@Nullable Integer runtimeOption,
+            @Nullable Integer newOption, @Nullable Integer oldOption, int defaultValue) {
+        if (runtimeOption != null) {
+            return runtimeOption;
+        }
+        return chooseBestValue(newOption, oldOption, defaultValue);
+    }
+
```
Comment (jeremyk-91) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java]: Hmm, this looks like `ObjectUtils.firstNonNull`, looking at this and the existing `chooseBestValue`? There might be a perf impact with some boxing/unboxing going on but it's probably very minor...

(I figure this and the older `chooseBestValue` is here to either give naming to the params or in case we want to change the logic in the future. In which case we can leave the methods as they are, but change the impl to simply hit `ObjectUtils.firstNonNull`)

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java [hunk 12]
```diff
@@ -258,9 +333,9 @@ private static SerializableTransactionManager create(
                 transactionManager,
                 kvs,
                 sweepRunner,
-                Suppliers.ofInstance(config.enableSweep()),
-                Suppliers.ofInstance(config.getSweepPauseMillis()),
-                Suppliers.ofInstance(getSweepBatchConfig(config)),
+                () -> MoreObjects.firstNonNull(runtimeConfig.get().enableSweep(), config.enableSweep()),
+                () -> MoreObjects.firstNonNull(runtimeConfig.get().getSweepPauseMillis(), config.getSweepPauseMillis()),
+                () -> getSweepBatchConfig(runtimeConfig.get(), config),
```
Comment (jeremyk-91) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java]: There is a race condition on the Sweep config here, in that in a single run of `grabLocksAndRun()` you could see `isSweepEnabled`, various `batchConfig`s and a `sweepPauseMillis` from different configs if the config changes are suitably timed.

This might well be benign though. We could fix this by supplying a config instead of the 3 things separately (and having the sweeper take this out and use it as a reference at the beginning of a single run), though that would rely on changes in `BackgroundSweeperImpl`. @gsheasby any thoughts on the safety of this?

**Comment by jeremyk-91** [comment]
> For a sample of what it takes to adapt this to existing products, check PR #2671 on internal core product.

Discussed offline and there was a manual test. My bad 😅

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java (hunk 7 above)
Comment (fsamuel-bs) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java]: We can't push them down to create, because we want to log specifically when:
- the product that uses the atlasdb library calls TM.create() with the new runtime config parameter.
- the runtime config is not present.
This log should target the person managing the stack, to make him/her move parameters from install configs to runtime configs.
We should have incentives other than logging to make products call the new TM.create(), one of them being marking the old calls as deprecated.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java (hunk 11 above)
Comment (fsamuel-bs) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java]: I actually didn't know this method existed. I was looking only on MoreUtils.firstNonNull. Thanks!

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java (hunk 12 above)
Comment (fsamuel-bs) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java]: While I understand that the runtime parameters may change in the same run, I agree that this is not a big problem.
If we were to create a config, we would probably pass it in as a parameter of `new BackgroundSweeperImpl()`. But modifying this call would be a breaking change, since there a other products who call it directly.
We could still create a config and 'snapshot' the parameters before each run, but this also seems like a stretch for its value.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java (hunk 10 above)
Comment (fsamuel-bs) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java]: The purpose if the log is to make people who manage a stack that supports runtime configs actually add them.
While the name of the method denotes that the runtime config will be validated, i.e. at each run will be validated and we'll abort if there's something invalid, this was not its purpose. I cannot think of a better name though.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java [hunk 13]
```diff
@@ -157,39 +163,50 @@ public long getSweepPersistentLockWaitMillis() {
...
     public abstract Integer getSweepDeleteBatchHint();
 
     /**
-     * @deprecated Use {@link #getSweepReadLimit()}, {@link #getSweepCandidateBatchHint()}
-     * and {@link #getSweepDeleteBatchHint()} instead.
+     * @deprecated Use {@link AtlasDbRuntimeConfig#getSweepReadLimit()},
+     * {@link AtlasDbRuntimeConfig#getSweepCandidateBatchHint()} and
+     * {@link AtlasDbRuntimeConfig#getSweepDeleteBatchHint()}.
      */
     @Deprecated
```
Comment (nziebart) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java]: sorry, i didn't understand why we can't get rid of these deprecated fields and make the others default. this is a new object - so we can just force people to use the new configs when they migrate to it.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbRuntimeConfig.java [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+    // TODO(ssouza): make this Default after removing support for mirror method on the AtlasDbConfig.
+    public abstract Boolean enableSweep();
+
+    /**
+     * The number of milliseconds to wait between each batch of cells
+     * processed by the background sweeper.
+     */
+    // TODO(ssouza): make this Default after removing support for mirror method on the AtlasDbConfig. We need to detect
+    // the absence of this value for now.
+    public abstract Long getSweepPauseMillis();
```
Comment (nziebart) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbRuntimeConfig.java]: can we move all the sweep paramters to a `SweepConfig` block within the `RuntimeConfig`? 

that would allow us to go ahead and make these fields default - the logic could be that if a `SweepConfig` is present in the runtime block, we will ignore any sweep config from the old atlas config (and we should throw if you have specified any sweep config there)

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java (hunk 12 above)
Comment (nziebart) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java]: This seems generally fine, but would be good to do a sanity check that this won't cause an exception somewhere (like if they disable sweep and null out some configs, we don't want a null pointer exception)

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java [hunk 15]
```diff
@@ -269,35 +345,44 @@ private static SerializableTransactionManager create(
         return transactionManager;
     }
 
-    private static SweepBatchConfig getSweepBatchConfig(AtlasDbConfig config) {
+    private static void validateRuntimeConfig(@Nullable AtlasDbRuntimeConfig runtimeConfig) {
+        if (runtimeConfig == null) {
+            log.warn("This service uses live reload configs. Please use the 'atlas-runtime' block for specifying"
```
Comment (nziebart) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java]: can we rephrase the first sentence to "AtlasDB now supports live reloadable configs."

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java (hunk 12 above)
Comment (fsamuel-bs) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java]: Right now, all suppliers calls go through the `getSweepBatchConfig` or `ObjectUtils.firstNonNull()`, which assures that no null value will be used. While this behavior seems correct now, I could see it being broken in the future. We should probably add a unit test to this.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbRuntimeConfig.java (hunk 14 above)
Comment (fsamuel-bs) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbRuntimeConfig.java]: While I like a SweepConfig block from an organizational standpoint (sweep configs would not be at root anymore, but at their own block, like the Leader config, or the TimeLockClient config), I feel this would make things more confusing for those migrating to runtime configs. We would have 5 possible supported configurations:
- AtlasConfig with no sweep configs -> currently resolving to default.
- AtlasConfig with custom sweep configs -> currently resolving custom sweep configs.
- AtlasRuntimeConfig with no SweepConfig -> would resolve to AtlasConfig configs.
- AtlasRuntimeConfig with SweepConfig -> would resolve to default SweepConfig values.
- AtlasRuntimeConfig with SweepConfig and custom values -> would resolve to custom  SweepConfig values.

Due to understandability, maybe we don't want to go down this route.

Comment (fsamuel-bs) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbRuntimeConfig.java]: Also, would this block be required?
- If not, after we delete the sweep configs from AtlasDB config, it's kind of redundant to have just the block or no block at all, since we would both use the default configs.
- If so, we probably just want to make it required it after we delete the original configs on the AtlasDB block. Otherwise, we could end up in a state where we require a new block even when supporting the old parameters.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java (hunk 12 above)
Comment (fsamuel-bs) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java]: This was addressed by adding a TODO on the AtlasDbConfig, to have add a default value on the new config when deleting the old config. The unit test doesn't seem valuable after we delete the old config. If people still think it would be good to add it, please let me know and I'll do so.

**Comment by fsamuel-bs** [comment]
I've addressed all the complaints on this PR. Note that we've decided to make this a devbreak change by deleting the old `TransactionManagers.create` methods without the AtlasDbRuntime parameter. We did this because we decided to group devbreak changes on the next release of AtlasDb, to avoid having to address them separately at each release.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java [hunk 16]
```diff
@@ -107,73 +110,89 @@ public static SerializableTransactionManager createInMemory(Schema schema) {
...
             AtlasDbConfig config,
+            java.util.function.Supplier<AtlasDbRuntimeConfig> runtimeConfig,
             Set<Schema> schemas,
             Environment env,
             LockServerOptions lockServerOptions,
             boolean allowHiddenTableAccess,
             String userAgent) {
+
+        java.util.function.Supplier<AtlasDbRuntimeConfig> wrappedRuntimeConfig = () -> {
+            validateRuntimeConfig(runtimeConfig.get());
```
Comment (fsamuel-bs) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java]: Agreed. The solutions that I see for this are:
- move it outside the supplier. This would make it log once, on startup, if the runtime config is not there.
- add some logic to prevent this from logging every single time, as in, just log once every 10m.

I will go for the first option, since it seems simpler and we don't want to overcomplicate this.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java [hunk 17]
```diff
@@ -107,73 +110,89 @@ public static SerializableTransactionManager createInMemory(Schema schema) {
...
             boolean allowHiddenTableAccess,
             Class<?> callingClass) {
-        return create(config, schemas, env, lockServerOptions, allowHiddenTableAccess,
+        return create(config, runtimeConfig, schemas, env, lockServerOptions, allowHiddenTableAccess,
                 UserAgents.fromClass(callingClass));
     }
 
     private static SerializableTransactionManager create(
             AtlasDbConfig config,
+            java.util.function.Supplier<AtlasDbRuntimeConfig> runtimeConfig,
```
Comment (fsamuel-bs) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java]: Agreed. `Supplier<Optional<RuntimeConfig>>` seems to express better what we expect from this parameter from an API standpoint. Note that there's a method for creating an `AtlasDbRuntimeConfig` from the `AtlasDbConfig` though, that could, in the future, return an empty config. But let's move forward with the optional parameter for now, and use such a method internally.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbRuntimeConfig.java [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+     */
+    // TODO(ssouza): make this Default after removing support for mirror method on the AtlasDbConfig. We need to detect
+    // the absence of this value for now.
+    @Nullable
+    public abstract Integer getSweepCandidateBatchHint();
+
+    /**
+     * The target number of (cell, timestamp) pairs to delete at once while sweeping.
+     */
+    // TODO(ssouza): make this Default after removing support for mirror method on the AtlasDbConfig. We need to detect
```
Comment (fsamuel-bs) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbRuntimeConfig.java]: Oh, I see your point now. And I agree that we shouldn't mix parameters from the AtlasDbConfig and the AtlasDbRuntimeConfig, as this change is currently doing.

But the problem with ignoring AtlasDbConfig parameters when there's a sweep block (or the presence of the `atlasdb-runtime` block) is that we would ignore custom parameters on the `atlasdb-config` block in preference of the default parameters of the AtlasDbRuntimeConfig, and this could be seen as a user-break change. We could detect this and log a warning though. @tboam, thoughts?

Comment (tboam) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbRuntimeConfig.java]: I share your concern about ignoring custom parameters in the old block and using defaults.  However I think a log message at `warn` in the start up logs is going to be missed by a lot of people.

Looking through this change there's some breaks to config already (apologies if this isn't the case, I've only skimmed the conversation) so I would go further and fail to start.  If you specify a sweep block but have static configuration for sweep too then don't start and tell people to move their config over.

I'd rather make some aggressive breaks now while this isn't used in many places than have to careful deprecate and remove once it is used in many places.  (You could even fail if there's any sweep config in the old block and require everyone to move the config block over now - move fast and break things!)

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java [hunk 19]
```diff
@@ -269,35 +289,45 @@ private static SerializableTransactionManager create(
         return transactionManager;
     }
 
-    private static SweepBatchConfig getSweepBatchConfig(AtlasDbConfig config) {
+    private static void validateRuntimeConfig(
+            java.util.function.Supplier<java.util.Optional<AtlasDbRuntimeConfig>> runtimeConfig) {
+        if (!runtimeConfig.get().isPresent()) {
+            log.warn("AtlasDB now supports live reloadable configs. Please use the 'atlas-runtime' block for specifying"
+                    + " live-reloadable atlas configs. Default to using configs specified on the 'atlas' block");
```
Comment (j-baker) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java]: drive-by: note that atlasdb doesn't know which config key each config appears in, so if anyone deviates this error message will be very confusing. The atlas codebase used to have a bunch of error messages which assumed atlas was inside gotham - which obviously after a while became false.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbRuntimeConfig.java [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+    public abstract Boolean enableSweep();
+
+    /**
+     * The number of milliseconds to wait between each batch of cells
+     * processed by the background sweeper.
+     */
+    // TODO(ssouza): make this Default after removing support for mirror method on the AtlasDbConfig. We need to detect
+    // the absence of this value for now.
+    @Nullable
+    public abstract Long getSweepPauseMillis();
```
Comment (j-baker) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbRuntimeConfig.java]: drive-by: From my internal projects I can suggest that this kind of config becomes sub-optimal as soon as something else becomes live reloadable - e.g. it's better to have

```yaml
atlas-runtime:
    sweep:
        enabled: true
        readLimit: 1234
        candidateBatchHint: 50
        deleteBatchHint: 100
    timestamp:
        batchTarget: 100
        requestLimitPerSecond: 200
```

rather than

```yaml
atlas-runtime:
    sweepEnabled: true
    sweepReadLimit: 1234
    sweepCandidateBatchHint: 50
    sweepDeleteBatchHint: 100
    timestampBatchTarget: 100
    timestampRequestLimitPerSecond: 200
```

- definitely don't view either of my comments as anything but random comments, just my first impression while reading it.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbRuntimeConfig.java (hunk 18 above)
Comment (nziebart) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbRuntimeConfig.java]: Agree with @tboam , we should fail to start if you have config in both places. Of course we can't fail if it gets live-reloaded, but at least we can throw at startup.

Don't think we can go so far as to require the runtime config - as that would be a config break :)

File: docs/source/release_notes/release-notes.rst [hunk 21]
```diff
@@ -63,6 +63,38 @@ develop
...
+             - `getSweepPauseMillis`
+             - `getSweepReadLimit`
+             - `getSweepCandidateBatchHint`
+             - `getSweepDeleteBatchHint`
+
+           The parameters are loaded and used at every sweep run. Please check :ref:`Sweep <sweep>` for information on how to use them.
+
+           (`Pull Request <https://github.com/palantir/atlasdb/pull/1976>`__)
+
+    *    - |userbreak| |deprecated|
```
Comment (gmaretic) [inline docs/source/release_notes/release-notes.rst]: You don't need the deprecated tag IMO

File: atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java [hunk 22]
```diff
@@ -219,16 +230,14 @@ protected final void check() {
...
-                    getSweepReadLimit() == null
-                            && getSweepCandidateBatchHint() == null
-                            && getSweepDeleteBatchHint() == null,
-                    "Your configuration mixes both the old and the new parameters"
-                            + " for setting sweep batch sizes. Please use 'sweepMaxCellTsPairsToExamine',"
-                            + " 'sweepCandidateBatchSize' and 'sweepDeleteBatchSize' instead of the deprecated"
-                            + " 'sweepBatchSize' and 'sweepCellBatchSize'.");
-        }
+
+        Preconditions.checkState(getSweepBatchSize() == null
```
Comment (gmaretic) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java]: A bit awkward to have these deprecated, then fail to start if they are used. That said, looks like too much lift to have a cleaner solution.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/config/SweepConfig.java [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+@Value.Immutable
+public class SweepConfig {
+    /**
+     * If true, a background thread will periodically delete cells that
+     * have been overwritten or deleted. This differs from scrubbing
+     * because it is an untargeted cleaning process that scans all data
+     * looking for cells to delete.
+     */
+    @Value.Default
+    public Boolean enableSweep() {
```
Comment (gmaretic) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/config/SweepConfig.java]: As @j-baker suggested, it's probably worth it to rename these, i.e., `enableSweep` -> `enabled`, etc.

File: docs/source/release_notes/release-notes.rst [hunk 24]
```diff
@@ -63,6 +63,38 @@ develop
...
+           - `getSweepCandidateBatchHint`
+           - `getSweepDeleteBatchHint`
+
+           If any of these options are specified in the install config, **AtlasDB will fail to start**. Please specify them on the runtime config.
+
+           (`Pull Request <https://github.com/palantir/atlasdb/pull/1976>`__)
+
+    *    - |devbreak|
+         - As part of the live-reloading work, were removed the methods `TransactionManagers.create()` without the AtlasDB runtime config.
+           If needed, the helper method `AtlasDbRuntimeConfig.create(AtlasDbConfig config)` can be used to create a runtime config from the AtlasDbConfig.
```
Comment (gmaretic) [inline docs/source/release_notes/release-notes.rst]: Was this since removed?

**Comment by fsamuel-bs** [comment]
Since we've released 0.45 last Friday, this should be ready for review once more.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java [hunk 25]
```diff
@@ -113,68 +113,77 @@ public static SerializableTransactionManager createInMemory(Schema schema) {
...
     public static SerializableTransactionManager create(
             AtlasDbConfig config,
+            java.util.function.Supplier<java.util.Optional<AtlasDbRuntimeConfig>> runtimeConfig,
             Set<Schema> schemas,
             Environment env,
             boolean allowHiddenTableAccess) {
-        log.info("Called TransactionManagers.create on thread {}. This should only happen once.",
-                Thread.currentThread().getName());
-        return create(config, schemas, env, LockServerOptions.DEFAULT, allowHiddenTableAccess);
+        log.info("Called TransactionManagers.create with live reloading config on thread {}."
```
Comment (hsaraogi) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java]: Remove the placeholder here. See #2074.

File: docs/source/release_notes/release-notes.rst [hunk 26]
```diff
@@ -65,6 +65,40 @@ develop
            (`Pull Request <https://github.com/palantir/atlasdb/pull/2060>`__)
            (`Pull Request <https://github.com/palantir/atlasdb/pull/2060>`__)
 
+         - Added support for live-reloading sweep configs.
+
+           There's a new set of `TransactionManagers.create()` methods, which accept a Supplier of the AtlasDbRuntimeConfig. The parameters of the new config are Sweep related:
+
+             - `enableSweep`
+             - `getSweepPauseMillis`
```
Comment (hsaraogi) [inline docs/source/release_notes/release-notes.rst]: config params should still be sweepPauseMillis etc.

File: docs/source/release_notes/release-notes.rst [hunk 27]
```diff
@@ -65,35 +65,37 @@ develop
...
+             - `getSweepCandidateBatchHint`
+             - `getSweepDeleteBatchHint`
 
            If any of these options are specified in the install config, **AtlasDB will fail to start**. Please specify them on the runtime config.
 
            (`Pull Request <https://github.com/palantir/atlasdb/pull/1976>`__)
 
     *    - |devbreak|
-         - As part of the live-reloading work, were removed the methods `TransactionManagers.create()` without the AtlasDB runtime config.
+         - As part of the live-reloading work, were removed the methods `TransactionManagers.create()` without the runtime config parameter.
```
Comment (gmaretic) [inline docs/source/release_notes/release-notes.rst]: we removed

Comment (fsamuel-bs) [inline docs/source/release_notes/release-notes.rst]: Rewrote the sentence to make more sense.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbRuntimeConfig.java (hunk 14 above)
Comment (gsheasby) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbRuntimeConfig.java]: The mythical 1.0 would be a good time to reduce this confusion - we could simpify the config at that point and have one place to add configs.
